server/sonar-server/src/main/java/org/sonar/server/computation/measure/Measure.java

```diff
@@ -137,37 +146,93 @@ public Builder forCharacteristic(int characteristicId) {
+
     public Measure create(boolean value, @Nullable String data) {
-      return new Measure(ValueType.BOOLEAN, ruleId, characteristicId, value ? 1.0d : 0.0d, data, null, description);
+      return new Measure(ValueType.BOOLEAN, ruleId, characteristicId, value ? 1.0d : 0.0d, data, null, description, qualityGateStatus, variations);
     }
 
     public Measure create(int value, @Nullable String data) {
-      return new Measure(ValueType.INT, ruleId, characteristicId, (double) value, data, null, description);
+      return new Measure(ValueType.INT, ruleId, characteristicId, (double) value, data, null, description, qualityGateStatus, variations);
     }
 
     public Measure create(long value, @Nullable String data) {
-      return new Measure(ValueType.LONG, ruleId, characteristicId, (double) value, data, null, description);
+      return new Measure(ValueType.LONG, ruleId, characteristicId, (double) value, data, null, description, qualityGateStatus, variations);
     }
 
     public Measure create(double value, @Nullable String data) {
-      return new Measure(ValueType.DOUBLE, ruleId, characteristicId, value, data, null, description);
+      return new Measure(ValueType.DOUBLE, ruleId, characteristicId, value, data, null, description, qualityGateStatus, variations);
     }
 
     public Measure create(String value) {
-      return new Measure(ValueType.STRING, ruleId, characteristicId, null, requireNonNull(value), null, description);
+      return new Measure(ValueType.STRING, ruleId, characteristicId, null, requireNonNull(value), null, description, qualityGateStatus, variations);
     }
 
     public Measure create(Level level) {
-      return new Measure(ValueType.LEVEL, ruleId, characteristicId, null, null, requireNonNull(level), description);
+      return new Measure(ValueType.LEVEL, ruleId, characteristicId, null, null, requireNonNull(level), description, qualityGateStatus, variations);
     }
 
     public Measure createNoValue() {
-      return new Measure(ValueType.NO_VALUE, ruleId, characteristicId, null, null, null, description);
+      return new Measure(ValueType.NO_VALUE, ruleId, characteristicId, null, null, null, description, qualityGateStatus, variations);
+    }
+  }
+
+  public static final class UpdateMeasureBuilder {
+    private final Measure source;
+    private QualityGateStatus qualityGateStatus;
+    private MeasureVariations variations;
+
+    public UpdateMeasureBuilder(Measure source) {
+      this.source = requireNonNull(source, "Can not create a measure from null");
+    }
+
+    /**
+     * Sets the QualityGateStatus of the updated Measure to create.
+     *
+     * @throws NullPointerException if the specified {@link QualityGateStatus} is {@code null}
+     * @throws UnsupportedOperationException if the source measure already has a {@link QualityGateStatus}
+     */
+    public UpdateMeasureBuilder setQualityGateStatus(QualityGateStatus qualityGateStatus) {
+      if (source.qualityGateStatus != null) {
```

Comment:
Because I think an `UnsupportedOperationException` is a more appropriate exception here. The state of the builder isn't incorrect since we refuse the change that would lead to it, the operation the consumer is trying to attemp is.